python/cudf/cudf/core/column/categorical.py

```diff
@@ -926,18 +932,20 @@ def __cuda_array_interface__(self):
         )
 
     def to_pandas(self, index=None):
-        codes = self.cat().codes.fillna(-1).to_array()
+        signed_dtype = min_scalar_type(len(self.categories))
+        codes = self.cat().codes.astype(signed_dtype).fillna(-1).to_array()
         categories = self.categories.to_pandas()
         data = pd.Categorical.from_codes(
             codes, categories=categories, ordered=self.ordered
         )
         return pd.Series(data, index=index)
 
     def to_arrow(self):
+        signed_codes_dtypes = min_scalar_type(len(self.categories))
         return pa.DictionaryArray.from_arrays(
             from_pandas=True,
             ordered=self.ordered,
-            indices=self.as_numerical.to_arrow(),
+            indices=self.as_numerical.astype(signed_codes_dtypes).to_arrow(),
             dictionary=self.categories.to_arrow(),
         )
```

Comment:
Does Arrow use signed types for DictionaryArray objects?

Reply:
Yes, when I tried to pass unsigned, it was crashing/failing.

Reply:
```
../src/arrow/type.cc:584:  Check failed: _s.ok() Operation failed: ValidateParameters(*index_type_, *value_type_)
Bad status: Type error: Dictionary index type should be signed integer, got uint8
/home/rgsl888/anaconda3/envs/cudf_dev2/lib/libarrow.so.17(+0x297744)[0x7fa7540d5744]

```

Reply:
Got it... We may want to reconsider this then. @davidwendt @jrhemstad what are our plans for libcudf DICTIONARY types with regards to signed vs unsigned integers?

Reply:
I had not given it much thought. Restricting to just signed (or just unsigned) for DICTIONARY indices would be helpful.

Reply:
Agreed, we should just pick signed or unsigned and stick with that. I'd lean towards unsigned so we get the widest range of values.

Reply:
Arrow-cpp not supporting unsigned types for dictionary indices seems like a bug we should file. It's supported in the other Arrow implementations.

Reply:
The max length for an Arrow Array is `(2^31)-1`, so I it makes sense they'd clamp to int32, but still seems like a bug not to allow smaller unsigned types.

Reply:
This seems to be as per their design https://arrow.apache.org/docs/format/Columnar.html#dictionary-encoded-layout
```
When a field is dictionary encoded, the values are represented by an array of signed integers representing the index of the value in the dictionary.
```

Reply:
@davidwendt Will you be moving forward with unsigned indices of Dictionary ?

Reply:
@davidwendt This is a blocker for https://github.com/rapidsai/cuml/pull/2267, is it possible to have an update on the question above?

Reply:
Whether the libcudf dictionary32 column should use unsigned vs signed types for indices should not be blocking anything here since it is not being  used by cudf at the moment as far as I know.

Reply:
Currently cudf Python uses signed types, this PR is proposing a change to use unsigned types, but the longer term plan is to push this functionality down to libcudf, via the dictionary32 work. We'd like to align behaviors as much as possible so as to not have users then rely on unsigned integer behaviors to then change it to signed integer behaviors when we make that change.

Reply:
Sounds good. I think we just need to create an issue to change dictionary32 to use unsigned types then. This work should not be required to unblock progress here though right?

Reply:
As you mentioned currently python is not dependent on libcudf, and if `dictionary32` can handle unsigned type indices in future, then this PR can go ahead.

Reply:
I have created bug requesting the change #5501, please feel free to update/discuss.

Reply:
Can someone send an email to the dev list or raise a JIRA ticket regarding unsigned types in dictionary columns in Arrow. Seems like their spec says the indices are signed, but @trxcllnt is claiming that implementations are showing otherwise.

Reply:
I had asked them about it in one of the PR which had that change and as per them it's for cross language support, https://github.com/apache/arrow/pull/6213#discussion_r442434501